### PR TITLE
feat:block_err add opts func ,add RegistryBlockType func;

### LINKED ISSUE #385: core.base package block_error add BlockErrorOption and RegistryBlockType for extend
## Issue Description

Type: *feature request*

### Describe what feature you want

feat: 
1. Add option func for block error to facilitate expansion, and add New and Reset methods;
2. Add custom registration BlockType function;

refactor:
1. Optimize performance of  blockType String function. detail info in bench function.

### PR DESCRIPTION
**Pull request by cafra**
### Describe what this PR does / why we need it

feat:

Add option func for block error to facilitate expansion, and add New and Reset methods;
Add custom registration BlockType function;

refactor:

Optimize performance of blockType String function. detail info in bench function.

### Does this pull request fix one issue?

 Fixes #385 

### Describe how you did it

feat:

Add option func for block error to facilitate expansion, and add New and Reset methods;
Add custom registration BlockType function;

refactor:

Optimize performance of blockType String function. detail info in bench function.

### Describe how to verify it

Provide unit testing and performance testing.

### Special notes for reviews

### CODE REVIEW

**Comment by sczyh30** [comment]
Refer https://github.com/alibaba/sentinel-golang/pull/388#issuecomment-770535489

File: core/base/result.go [hunk 1]
```diff
@@ -28,25 +29,38 @@ const (
 	BlockTypeCircuitBreaking
 	BlockTypeSystemFlow
 	BlockTypeHotSpotParamFlow
+	// BlockTypeRegistryStart convenient to expand the type to avoid conflicts
+	BlockTypeRegistryStart
 )
 
+var (
+	blockTypeNames = []string{
```
Comment (sanxun0325) [inline core/base/result.go:37]: Use Map to store types and descriptions and cancel BlockTypenames. Is that OK?

Comment (louyuting) [inline core/base/result.go:37]: Use slice does has the better performance, that's good!
But i have some concern about the maintainability, the order of const variable can not be changed, this is where potential bugs can arise.

File: core/base/result.go [hunk 2]
```diff
@@ -28,25 +29,38 @@ const (
...
+		BlockTypeUnknown:          "Unknown",
+		BlockTypeFlow:             "FlowControl",
+		BlockTypeIsolation:        "BlockTypeIsolation",
```
[the remark this answers is not in the document]
Comment (cafra) [reply core/base/result.go:40]: image
Keep the value consistent with the previous version. If the solution needs to be consistent, add the prefix BlockType uniformly.

File: core/base/result.go (hunk 1 above)
Comment (cafra) [reply core/base/result.go:37]: image

The performance of map is poor. 
If you feel that the code is inconvenient to maintain, 
modify the version of the "switch",is it ok ?

Comment (cafra) [reply core/base/result.go:37]: OK。In order to improve the scalability and facilitate the registration of new types, it is modified to map. Although the performance of map is poor, it is acceptable.

image

Comment (cafra) [reply core/base/result.go:37]: In order to improve the scalability and facilitate the registration of new types, it is modified to map. Although the performance of map is poor, it is acceptable.

File: core/base/result.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cafra) [reply core/base/result.go:40]: image

The value does not need to be consistent, so the string length is too long and the log print is too long

File: core/base/result.go (hunk 1 above)
Comment (louyuting) [reply core/base/result.go:37]: Based on your bench, the performance of map is bad.
It's not suggested.

File: core/base/block_error_test.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+			}},
+			want: NewBlockError(WithBlockType(BlockTypeFlow),
+				WithBlockMsg("test"),
+				WithRule(new(MockRule)),
+				WithSnapshotValue("snapshot")),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := NewBlockError(tt.args.opts...); !reflect.DeepEqual(got, tt.want) {
```
[the remark this answers is not in the document]
Comment (cafra) [reply core/base/block_error_test.go:33]: ok.

image

**Review by louyuting — APPROVED** [review]
LGTM

**Review by sanxun0325 — APPROVED** [review]
LGTM

File: core/base/result.go (hunk 1 above)
Comment (louyuting) [reply core/base/result.go]: Think of maintainability, map is acceptable.

**Pull request merged**